yarn-project/aztec-rpc/src/account_state/account_state.ts

```diff
@@ -51,45 +55,45 @@ export class AccountState {
-      throw new Error('Unknown contract.');
-    }
+  public async simulate(txRequest: TxRequest, contractDataOracle?: ContractDataOracle) {
+    // TODO - Pause syncing while simulating.
 
-    const selector = txRequest.functionData.functionSelector;
-    const functionDao = contract.functions.find(f => f.selector.equals(selector));
-    if (!functionDao) {
-      throw new Error('Unknown function.');
+    if (!contractDataOracle) {
+      contractDataOracle = new ContractDataOracle(this.db, this.node);
     }
 
-    const oldRoots = new OldTreeRoots(Fr.ZERO, Fr.ZERO, Fr.ZERO, Fr.ZERO); // TODO - get old roots from the database/node
-
-    // TODO - Pause syncing while simulating.
-    this.log(`Executing simulator...`);
-    const executionResult = await this.simulator.run(
-      txRequest,
-      functionDao,
+    const contractAddress = txRequest.to;
+    const functionAbi = await contractDataOracle.getFunctionAbi(
       contractAddress,
-      contract.portalContract,
-      oldRoots,
+      txRequest.functionData.functionSelector,
     );
+    const portalContract = await contractDataOracle.getPortalContractAddress(contractAddress);
+    const oldRoots = new OldTreeRoots(Fr.ZERO, Fr.ZERO, Fr.ZERO, Fr.ZERO); // TODO - get old roots from the database/node
 
-    return { contract, oldRoots, executionResult };
+    const simulatorOracle = new SimulatorOracle(contractDataOracle, this.db, this.keyPair);
+    const simulator = new AcirSimulator(simulatorOracle);
+    this.log('Executing simulator...');
+    const result = await simulator.run(txRequest, functionAbi, contractAddress, portalContract, oldRoots);
+    this.log('Simulation completed!');
+
+    return result;
   }
 
-  public createUnverifiedData(contract: AztecAddress, newNotes: { preimage: Fr[]; storageSlot: Fr }[]): UnverifiedData {
-    const txAuxDatas = newNotes.map(({ preimage, storageSlot }) => {
-      const notePreimage = new NotePreimage(preimage);
-      return new TxAuxData(notePreimage, contract, storageSlot);
-    });
-    const chunks = txAuxDatas.map(txAuxData => {
-      // TODO - Should use the correct recipient public key.
-      const recipient = this.publicKey;
-      return txAuxData.toEncryptedBuffer(recipient, this.grumpkin);
-    });
-    return new UnverifiedData(chunks);
+  public async simulateAndProve(txRequest: TxRequest, signature: EcdsaSignature) {
+    // TODO - Pause syncing while simulating.
+
+    const contractDataOracle = new ContractDataOracle(this.db, this.node);
+    const executionResult = await this.simulate(txRequest, contractDataOracle);
+
+    const kernelProver = new KernelProver(contractDataOracle);
+    this.log('Executing Prover...');
+    const { proof, publicInputs } = await kernelProver.prove(txRequest, signature, executionResult);
+    this.log('Proof completed!');
+
+    const contractAddress = txRequest.to;
+    const unverifiedData = this.createUnverifiedData(contractAddress, executionResult.preimages.newNotes);
```

Comment:
We need to call this recursively for all the execution graph! It's not accumulated in the top level to be able to know which step of the execution / contract generated which notes :D

Reply:
Right! Good spot :)